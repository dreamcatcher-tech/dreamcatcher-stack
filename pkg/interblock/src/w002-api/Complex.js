import assert from 'assert-fast'
export default class Complex {
  #parent
  static create({ state, binary, network, actions, isLoading, wd, tree }) {
    return new Complex({ state, binary, network, actions, isLoading, wd, tree })
  }
  constructor({ state, binary, network, actions, isLoading, wd, tree }) {
    this.state = state || {}
    this.binary = binary
    this.network = network || []
    this.actions = actions || {}
    this.isLoading = isLoading || false
    this.wd = wd || '/'
    this.tree = tree
    if (tree === undefined) {
      this.tree = Complex.create({ ...this, tree: false })
    }
  }
  clone() {
    const next = new Complex(this)
    next.#parent = this.#parent
    return next
  }
  addAction(actions) {
    const next = this.clone()
    next.actions = { ...next.actions, ...actions }
    return next
  }
  setWd(wd) {
    assert.strictEqual(typeof wd, 'string')
    assert(wd.startsWith('/'))
    const next = this.clone()
    next.wd = wd
    return next
  }
  child(path) {
    assert.strictEqual(typeof path, 'string')
    assert(!path.startsWith('/'))
    assert(!path.startsWith('./'))
    assert(!path.startsWith('..'))
    const child = this.network.find(({ path: p }) => p === path)
    if (!child) {
      throw new Error(`child not found: ${path}`)
    }
    const { tree } = this
    const next = Complex.create({ ...child, tree })
    next.#parent = this
    return next
  }
  hasChild(path) {
    if (!path) {
      return false
    }
    assert.strictEqual(typeof path, 'string')
    assert(!path.startsWith('/'))
    assert(!path.startsWith('./'))
    assert(!path.startsWith('..'))
    return this.network.some(({ path: p }) => p === path)
  }
  rm(path) {
    assert(this.hasChild(path))
    const next = this.clone()
    next.network = next.network.filter(({ path: p }) => p !== path)
    return next
  }
  setNetwork(network) {
    assert(Array.isArray(network))
    const next = this.clone()
    next.network = network
    return next
  }
  setState(state) {
    assert.strictEqual(typeof state, 'object')
    const next = this.clone()
    next.state = state
    return next
  }
  parent() {
    return this.#parent
  }
}
