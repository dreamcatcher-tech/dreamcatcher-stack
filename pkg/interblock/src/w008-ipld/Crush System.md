# The Crush System

This collection of classes are designed to break a Pulse down into IPFS compatible chunks, for efficient hashing.

## Requirements

1. use the inheritance of IpldInterface as identifying feature of classes that get crushed
1. crush relies on an unbroken chain of crushable classes from the root down
   1. ie: cannot have a pojo that holds a crush class in it and expect it to work
1. arrays of crush classes follow these rules
1. when flagged, some children may be 'punched out' as CID links, and reinflated back in
1. any tree of classes can be reinflated from ipfs back into classes
1. ipfs blocks representing the different between the previous 'crush' process can be fetched then persisted in ipfs storage
1. be able to directly represent an IPFS primitive as a class, such as a CID - these are needed to signal to the ORM that it should stop rehydration
1. Be able to connect a set of IPLD schemas to the classes, to check them for correctness on crush and uncrush
   1. This check should be disablable for speed
