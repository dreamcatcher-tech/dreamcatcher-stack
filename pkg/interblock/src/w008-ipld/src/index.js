export { Binary } from './Binary'
export { RequestId } from './RequestId'
export { Request } from './Request'
export { RxRequest } from './RxRequest'
export { Reply } from './Reply'
export { RxReply } from './RxReply'
export { Address } from './Address'
export { PulseLink, HistoricalPulseLink } from './PulseLink'
export { TxQueue } from './TxQueue'
export { Tx } from './Tx'
export { RxQueue } from './RxQueue'
export { Rx } from './Rx'
export { Channel } from './Channel'
export { Config } from './Config'
export { PublicKey } from './PublicKey'
export { Validators } from './Validators'
export { ChannelsHamt } from './ChannelsHamt'
export { ChildrenHamt } from './ChildrenHamt'
export { UplinksHamt } from './UplinksHamt'
export { DownlinksHamt } from './DownlinksHamt'
export { SymlinksHamt } from './SymlinksHamt'
export { AddressesHamt } from './AddressesHamt'
export { Channels } from './Channels'
export { Transmissions } from './Transmissions'
export { Network } from './Network'
export { AsyncRequest } from './AsyncRequest'
export { AsyncTrail } from './AsyncTrail'
export { Pending } from './Pending'
export { State } from './State'
export { Timestamp } from './Timestamp'
export { Dmz } from './Dmz'
export { Provenance } from './Provenance'
export { Interpulse } from './Interpulse'
export { Pulse } from './Pulse'
export { Keypair } from './Keypair'
export { decode } from './IpldUtils'
export { IpldInterface } from './IpldInterface'
export { Hamt } from './Hamt'
