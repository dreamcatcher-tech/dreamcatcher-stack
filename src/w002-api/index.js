const api = require('./src/api')
const hooks = require('./src/hooks')
const queries = require('./src/queries')
module.exports = { ...api, ...hooks, ...queries }
