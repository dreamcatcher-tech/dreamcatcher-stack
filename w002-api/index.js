const api = require('./src/api')
const promises = require('./src/promises')
module.exports = { ...api, ...promises }
