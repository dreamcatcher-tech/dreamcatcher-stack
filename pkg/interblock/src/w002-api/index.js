import assert from 'assert-fast'
import { Request } from '../w008-ipld'

export const interchain = (type, payload, to) => {
  const { interchain } = globalThis[Symbol.for('interblock:api:hook')]
  assert(interchain, `global hook detached`)
  return interchain(type, payload, to)
}
export const useState = (path = '.') => {
  const { useState } = globalThis[Symbol.for('interblock:api:hook')]
  assert(useState, `global hook detached`)
  return useState(path)
}
export const usePulse = async (path = '.') => {
  assert.strictEqual(typeof path, 'string')
  assert(path)
  const reply = interchain('@@USE_PULSE', { path })
  return reply
}

export const isApiAction = (request, covenant) => {
  assert.strictEqual(typeof request, 'object')
  assert.strictEqual(typeof request.type, 'string')
  assert(request.type)
  assert.strictEqual(typeof covenant, 'object')
  assert.strictEqual(typeof covenant.api, 'object')
  const schemas = Object.values(covenant.api)
  assert(schemas.length, 'No actions in schema')
  assert(schemas.every((schema) => typeof schema.title === 'string'))
  return schemas.some(({ title }) => request.type === title)
}

export const ensureChild = async (path, installer = 'unity') => {
  assert.strictEqual(typeof path, 'string')
  assert(path)
  // TODO assert child points to a deeper path, not higher one
  if (typeof installer === 'string') {
    installer = { covenant: installer }
  }
  assert.strictEqual(typeof installer, 'object')
  try {
    await interchain(Request.tryPath(path))
    // TODO assert covenant matches
  } catch (error) {
    // TODO work with any child
    const msg = `Segment not present: /.mtab of: .mtab`
    if (error.message !== msg) {
      throw error
    }
    const spawnAction = Request.createSpawn(path, installer)
    await interchain(spawnAction)
  }
}

export const isSystemAction = (request) => {
  assert.strictEqual(typeof request.type, 'string')
  return Request.SYSTEM_TYPES.includes(request.type)
}

export { schemaToFunctions } from './schemaToFunctions'
