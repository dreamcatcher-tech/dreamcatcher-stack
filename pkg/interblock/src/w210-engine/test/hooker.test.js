import { assert } from 'chai/index.mjs'
import { interchain } from '../../w002-api'
import { Engine } from '..'
import { Request } from '../../w008-ipld'
import Debug from 'debug'
const debug = Debug('interblock:tests:hooker')

describe('hooker', () => {
  test('loopback cleared immediately', async () => {
    const engine = await Engine.createCI()
    engine.overload({
      root: {
        reducer: async (request) => {
          if (request.type === 'TEST') {
            await interchain(Request.createSpawn('loop'))
            const result = await interchain('PING')
            return result
          }
          if (request.type === 'PING') {
            return { pong: true }
          }
        },
      },
    })
    const ping = Request.create({ type: 'TEST' })
    const result = await engine.pierce(ping)
    assert.strictEqual(result.pong, true)
    const pulse = engine.latest
    const loopback = await pulse.getNetwork().getLoopback()
    assert(loopback.rx.isEmpty())
    assert(loopback.tx.isEmpty())
    assert(loopback.rx.isSettled())
    assert(loopback.tx.isSettled())
    expect(loopback).toMatchSnapshot()
  })
  test.todo('throw if pending and tx a request to self')
  // if only request from pending is a request to self, then know it will never resolve
  // basically cannot raise pending, then request something to self
  test.todo('wait for all promises')
  test.todo('awaiting on self alone rejects')
  // set up a reducer that gets tampered with after promise resolves
  test.todo('covenant must make requests in same order')
})
