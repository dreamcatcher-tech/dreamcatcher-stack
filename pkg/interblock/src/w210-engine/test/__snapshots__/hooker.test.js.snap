// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hooker loopback cleared immediately 1`] = `
Channel {
  "address": Address {},
  "aliases": [
    ".",
  ],
  "channelId": 0,
  "rx": Rx {
    "reducer": RxQueue {
      "promisedReplies": [],
      "promisedRequestIds": [],
      "replies": [],
      "repliesLength": 1,
      "requests": [],
      "requestsLength": 1,
    },
    "system": RxQueue {
      "promisedReplies": [],
      "promisedRequestIds": [],
      "replies": [],
      "repliesLength": 2,
      "requests": [],
      "requestsLength": 2,
    },
    "tip": HistoricalPulseLink {},
  },
  "tx": Tx {
    "precedent": HistoricalPulseLink {},
    "reducer": TxQueue {
      "promisedReplies": [],
      "promisedRequestIds": [],
      "replies": [],
      "repliesLength": 1,
      "requests": [],
      "requestsLength": 1,
    },
    "system": TxQueue {
      "promisedReplies": [],
      "promisedRequestIds": [],
      "replies": [],
      "repliesLength": 2,
      "requests": [],
      "requestsLength": 2,
    },
  },
}
`;
